clojure-ts-mode.el

```diff
@@ -1870,10 +1876,119 @@ With universal argument \\[universal-argument], fully unwinds thread."
   (interactive)
   (clojure-ts-unwind '(4)))
 
+(defun clojure-ts--remove-superfluous-parens ()
+  "Remove extra parens from a form."
+  (when-let* ((node (treesit-thing-at-point 'sexp 'nested))
+              ((clojure-ts--list-node-p node))
+              ((= 1 (treesit-node-child-count node t))))
+    (let ((delete-pair-blink-delay 0))
+      (delete-pair))))
+
+(defun clojure-ts--thread-first ()
+  "Thread a sexp using ->."
+  (save-excursion
+    (clojure-ts--skip-first-child (clojure-ts--threading-sexp-node))
+    (down-list)
+    (treesit-beginning-of-thing 'sexp -1)
+    (let ((contents (clojure-ts--delete-and-extract-sexp)))
+      (delete-char -1)
+      (when (looking-at-p " *\n")
+        (join-line 'following))
+      (backward-up-list)
+      (insert contents "\n")
+      (clojure-ts--remove-superfluous-parens))))
+
+(defun clojure-ts--thread-last ()
+  "Thread a sexp using ->>."
+  (save-excursion
+    (clojure-ts--skip-first-child (clojure-ts--threading-sexp-node))
+    (treesit-end-of-thing 'sexp)
+    (down-list -1)
+    (treesit-beginning-of-thing 'sexp)
+    (let ((contents (clojure-ts--delete-and-extract-sexp)))
+      (delete-char -1)
+      (treesit-end-of-thing 'sexp -1 'restricted)
+      (when (looking-at-p " *\n")
+        (join-line 'following))
+      (backward-up-list)
+      (insert contents "\n")
+      (clojure-ts--remove-superfluous-parens))))
+
+(defun clojure-ts--threadable-p (node)
+  "Return non-nil if expression NODE can be threaded.
+
+First argument after threading symbol itself should be a list and it
+should have more than one named child."
+  (let ((second-child (treesit-node-child node 1 t)))
+    (and (clojure-ts--list-node-p second-child)
+         (> (treesit-node-child-count second-child t) 1))))
+
+(defun clojure-ts-thread ()
```

Comment:
Shouldn't this raise some error if the form at point is not threadable?

Reply:
In `clojure-ts-mode` it just returns `nil`, I did the same here. Maybe because it's used in a loop in both `thread-first-all` and `thread-last-all`?

Reply:
Much of the refactoring code was copied from `clj-refactor` in bulk early on, so I didn't pay close attention to each command there - that's why I notice some stuff for the first time.

I think it probably makes sense to have an optional param controlling whether to raise an error or return `nil`, and perhaps even introduce some lower-level function that's not actually command that all of the commands call. (extracted from `clojure-ts-thread`)

Any rate - what you're doing right now is a great opportunity for us to examine closely each refactoring command and make them slightly better here and there in the process of porting them to `clojure-ts-mode`.

Reply:
> I think it probably makes sense to have an optional param controlling whether to raise an error or return `nil`, and perhaps even introduce some lower-level function that's not actually command that all of the commands call. (extracted from `clojure-ts-thread`)

Perhaps we could look into it when we port more refactoring commands and see some pattern that can be extracted.